daml/ContingentClaims/Lifecycle.daml

```diff
@@ -78,20 +78,26 @@ lifecycle' _ _ (OneF id) = do
   pure $ One id
 lifecycle' _ _ (GiveF (_, m)) = local negate m
 lifecycle' spot _ (WhenF obs (c, m)) = do
-  wasLifecycled <- lift . lift $ eval spot obs
-  if wasLifecycled then m $> Zero else pure $ When obs c
+  predicate <- lift . lift $ eval spot obs
+  if predicate then m $> Zero else pure $ When obs c
 lifecycle' spot _ (ScaleF obs (_, m)) = do
   k <- lift . lift $ eval spot obs
-  local (* k) m
+  Scale obs <$> local (* k) m
 lifecycle' _ choose (OrF (c, m) (c', m')) = do
   let chosen = choose c c'
   if | chosen == c -> m
      | chosen == c' -> m'
      | otherwise -> abort "lifecycle: Invalid branch election"
-lifecycle' _ _ (AndF (_, m) (_, m')) = reduceZero <$> m <*> m' where
-   reduceZero Zero c' = c'
-   reduceZero c Zero = c
-   reduceZero c c' = And c c'
+lifecycle' _ _ (AndF (_, m) (_, m')) = And <$> m <*> m'
 lifecycle' spot _ (CondF obs (_, m) (_, m')) = do
   predicate <- lift . lift $ eval spot obs
   if predicate then m else m'
```

Comment:
I think this needs to be manually pruned here; one of the `Cond` branches will be lifecycled, so we can remove the whole subtree. Either here, or probably better in `pruneZeros'`. It's just a bit confusing as we could have `Cond p (One "USD") (One "EUR")`, in which case, neither branch is `Zero`, but this would still get lifecycled, so we need to prune it.

Reply:
Resolved with the new top-down refactor.